repo: remotes family: implement list/get-url/set-url

jtagcat opened the pull request:
edit: ready to review

jtagcat commented:
Are style for func names enforced? It'd make sense to use RemotesList instead of ListRemotes; the same with (upcoming PR) RemoteURLGet and RemoteURLSet (I prefer) vs GetRemoteURL and SetRemoteURL (following the style in the file).

I'd rather rename ListRemotes() to RemotesList()

Why?
 - Autocomplete
 - Hierarchy
 - Following upstream (git bin)

jtagcat commented:
force-push: moved a part to utils.go; reusing it in upcoming code

jtagcat commented:
@unknwon PTAL, planning to do similar PRs for additional commands.

unknwon commented:
> It'd make sense to use RemotesList instead of ListRemotes; the same with RemoteURLGet and RemoteURLSet (used here) vs GetRemoteURL and SetRemoteURL (style in file).
> 
> Are style for func names enforced? I didn't follow the exact naming in the same file.

No there is no existing enforced naming convention, but I think your idea of being closer to how a user would use Git bin makes total sense 👍

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+// RemoteURLGetOptions contains arguments for retrieving URL(s) of a remote of the repository.
+// Docs: https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emget-urlem
+type RemoteURLGetOptions struct {
+	// False: get fetch URLs
+	// True: get push URLs
+	Push bool
```
Godoc converts the line breaks into spaces, so this is how it looks like in hover-over:
<img width="281" alt="CleanShot 2022-02-10 at 12 44 53@2x" src="https://user-images.githubusercontent.com/2946214/153338769-85077f4c-4ceb-4470-baa3-a045512c61c7.png">

```suggestion
	// Indicates whether to get push URLs instead of fetch URLs.
	Push bool
```

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+	// False: get fetch URLs
+	// True: get push URLs
+	Push bool
+	// True: get all URLs (lists also non-main URLs; not related with Push)
+	All bool
```
```suggestion
	// Indicates whether to get all URLs, including lists that are not part of main URLs. This option is independent of the Push option.
	All bool
```

unknwon replied:
This comment seems overlooked.

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+// RemoteURLSetOptions contains arguments for setting an URL of a remote of the repository.
+// Docs: https://git-scm.com/docs/git-remote#Documentation/git-remote.txt-emget-urlem
+type RemoteURLSetOptions struct {
+	// False: set fetch URLs
+	// True: set push URLs
+	Push bool
```
```suggestion
	// Indicates whether to get push URLs instead of fetch URLs.
	Push bool
```

unknwon replied:
This comment seems overlooked.

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+	return RepoRemoteURLAdd(r.path, name, newvalue, opts...)
+}
+
+// RepoRemoteURLDelRegex Deletes all URLs matchin regex of the remote with given name of the repository in given path.
```
```suggestion
// RepoRemoteURLDelRegex deletes all URLs matching regex of the remote with given name of the repository in given path.
```

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+	return err
+}
+
+// RemoteURLDelRegex // RepoRemoteURLDelRegex Deletes all URLs matchin regex of the remote with given name of the repository.
```
```suggestion
// RemoteURLDelRegex deletes all URLs matching regex of the remote with given name of the repository.
```

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+}
+
+// RepoRemotesList lists remotes of the repository in given path.
+func RepoRemotesList(repoPath string, opts ...RemotesListOptions) ([]string, error) {
```
If we're following how we use git bin, this should be like this? (if so, same for other methods and option structs)
```suggestion
func RepoRemotes(repoPath string, opts ...RemotesListOptions) ([]string, error) {
```

jtagcat replied:
Renamed it, though kept `RemotesOptions` might get confusing?

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+	Timeout time.Duration
+}
+
+// RepoRemoteURLSetFirst sets first URL of the remote with given name of the repository in given path.
+func RepoRemoteURLSetFirst(repoPath, name, newurl string, opts ...RemoteURLSetOptions) error {
```
How is the "first" come into play? Not seeing "first" is used as git args?

jtagcat replied:
uhh yeah so.. repo can have like
origin urls:
 - type fetch url: foobar1
 - type push url: foobar2
 - type fetch url: foobar3
 - type fetch url: foobar4
 
^^^ in this case, the first fetch url is foobar1, and first push url is foobar3

In reality, this almost never happens, and even if it does, the 'first' is used elsewhere as well.

It might be a good idea to add a short 'it is almost always first, and if not, you probably don't have to worry', but don't know on how to word it.

unknwon replied:
Thanks for the context!

Either way I think we do not need explicitly put "first" in our function name because we are basically pass-on the value from Git CLI, and whatever it returns is what we return to the caller.

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,181 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+}
+
+// RepoRemoteURLSetRegex sets the first URL of the remote with given name and URL regex match of the repository in given path.
+func RepoRemoteURLSetRegex(repoPath, name, urlregex string, newurl string, opts ...RemoteURLSetOptions) error {
```
```suggestion
func RepoRemoteSetURL(repoPath, name, urlregex string, newurl string, opts ...RemoteURLSetOptions) error {
```

We should name it like this to follow how we use git bin?

jtagcat replied:
regex: I decided in the past to split 'first' and regex for ease of use and clarity.

Alternatively: if `urlregex == ""`, could be a thing, but it implies needing an urlregex param (hope people read the comments). Throwing it under options doesn't sit well with me either.

***
> We should name it like this to follow how we use git bin?
 
In my initial suggestion, I was aiming for hierarchy, and set/get as an operation (`repo/remotes/URLs/set,setregex,get`).

Though, I'm also fine for opting to follow git bin. LMK if you'd like that.

unknwon replied:
> regex: I decided to split 'first' and regex for ease of use and clarity.
> 
> (Alternatively: if `urlregex == ""`, could be a thing, but it implies needing an `urlregex` param (hope people read the docs?). Throwing it under options doesn't sit well with me either.

Ah I see, did not realize this, I guess I'm OK with the "Regex" in the function name then.

> > We should name it like this to follow how we use git bin?
> 
> In my initial suggestion, I was aiming for hierarchy, and set/get as an operation (`repo/remotes/URLs/set,setregex,get`).
> 
> Though, I'm also fine for opting to follow git bin. LMK if you'd like that.

Let's all-in to follow how we use git bin!

unknwon replied:
After opening up the code locally, I think it makes more sense to make the URL regex be part of the options, as it is an option after all.

unknwon commented on utils.go:
```diff
@@ -71,3 +72,12 @@ func concatenateError(err error, stderr string) error {
 	return fmt.Errorf("%v - %s", err, stderr)
 }
+
+// turns "a\n\list\n" to []string{"a", "list"}
+func stdoutToStringSlice(stdout []byte) []string {
```
```suggestion
// bytesToStrings splits given bytes into strings by line separator ("\n"). It returns empty slice if the given bytes only contains line separators.
func bytesToStrings(in []byte) []string {
```

unknwon commented on utils.go:
```diff
@@ -71,3 +72,12 @@ func concatenateError(err error, stderr string) error {
+
+// turns "a\n\list\n" to []string{"a", "list"}
+func stdoutToStringSlice(stdout []byte) []string {
+	o := strings.TrimRight(string(stdout), "\n")
+	if o == "" { // empty (not {""}, len=1)
+		return []string{}
+	}
+	return strings.Split(o, "\n")
```
```suggestion
	s := strings.TrimRight(string(in), "\n")
	if s == "" { // empty (not {""}, len=1)
		return []string{}
	}
	return strings.Split(s, "\n")
```

unknwon reviewed (changes requested):
Thanks for the PR! Overall looking good 🤩 Mostly convention issues.

jtagcat commented:
Not squashing/amending previous commit atm to make it more readable for you, commits can be squashed later.

unknwon reviewed (commented):
Sorry about delayed response 😅

jtagcat commented:
> `All bool` https://github.com/gogs/git-module/pull/67#discussion_r812639561

I can't see how, but I improved the wording nevertheless.

edit: `True: ` part probably, removed it.

> `Push bool` https://github.com/gogs/git-module/pull/67#discussion_r812639696

fixed

> Either way I think we do not need explicitly put "first" in our function name because we are basically pass-on the value from Git CLI, and whatever it returns is what we return to the caller.

s/SetFirst/Set/g done

> Let's all-in to follow how we use git bin!

s/URLGet/GetURL/g, s/URLSet/SetURL/g

***

also jfyi you can make direct edits in the PR as you see fit

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,179 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+type RemoteGetURLOptions struct {
+	// Indicates whether to get push URLs instead of fetch URLs.
+	Push bool
+	// get all URLs: also lists non-main/default URLs for fetch OR push (not related with Push)
```
```suggestion
	// Indicates whether to get all URLs, including lists that are not part of main URLs. This option is independent of the Push option.
```

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,179 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+	return RepoRemoteSetURLRegex(r.path, name, urlregex, newurl, opts...)
+}
+
+// RepoRemoteURLAdd adds an URL to the remote with given name of the repository in given path.
+func RepoRemoteURLAdd(repoPath, name, newurl string, opts ...RemoteSetURLOptions) error {
+	var opt RemoteSetURLOptions
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+
+	cmd := NewCommand("remote", "set-url", "--add")
+	if opt.Push {
+		cmd.AddArgs("--push")
+	}
+
+	_, err := cmd.AddArgs(name, newurl).RunInDirWithTimeout(opt.Timeout, repoPath)
+	return err
+}
```
Any reason `--add` is not a field (`Add bool` )in `RemoteSetURLOptions` but an independent function? I'm going to make a push to merge these two functions.

jtagcat replied:
The logic behind here was that it is commonly used, and useful to separate it, but actually isn't on second thought, good catch.

jtagcat replied:
(they are separated in manpages asw)

![image](https://user-images.githubusercontent.com/38327267/155691194-d2affda3-2f0b-4e0e-a833-30a83c64ad03.png)

jtagcat replied:
Add does not take any regex.

jtagcat replied:
Generics might work here: allowing two different RemoteSetURLOptions.

Runtime checking or even git failing doesn't seem ok for me.

unknwon replied:
I see, I missed the part it does not accept the regex. Agree to have its own function then 😅

unknwon reviewed (approved):
Thanks again for the follow up!

unknwon commented:
I made a push in https://github.com/gogs/git-module/pull/67/commits/1e5f3c2bcf7372b64c99e62a974e909a9d73cbe2, LMK if you have any questions, otherwise I'm going to merge once looks good to you!

jtagcat commented:
![image](https://user-images.githubusercontent.com/38327267/155690580-65c743ff-72db-4b7c-b577-60619796b3ad.png)

Made a small change for the comment here, otherwise lgtm.

jtagcat commented:
RemoteSetURLDelete seems odd for me, RemoteDeleteURL is better imo

unknwon commented on repo_remote.go:
```diff
@@ -285,10 +285,10 @@ type RemoteSetURLDeleteOptions struct {
-func RemoteSetURLDelete(repoPath, name string, opts ...RemoteSetURLDeleteOptions) error {
-	var opt RemoteSetURLDeleteOptions
+func RemoteDeleteURL(repoPath, name string, opts ...RemoteDeleteURLOptions) error {
+	var opt RemoteDeleteURLOptions
```
`RemoteSetURLDelete` was used to be aligned with how Git bin is used, as `"remote", "set-url", "--delete"`. If we skip the `"set-url"` part, then I think we are not following how a user uses Git bin anymore, which kinda breaks the naming convention we try to develop since this PR.

jtagcat replied:
ok, reverting it

jtagcat commented:
suggested refactor out of this PR:
```go
type RemotesOptions struct {
	// The timeout duration before giving up for each shell command execution.
	// The default timeout duration will be used when not supplied.
	Timeout time.Duration
}
```
You can just do
```go
type DefaultOptions struct {
	// The timeout duration before giving up for each shell command execution.
	// The default timeout duration will be used when not supplied.
	Timeout time.Duration
}

// and for each:
var RemotesOptions DefaultOptions
```

unknwon commented:
> > How will `RemotesOptions` be used look like?
> 
> It will be the same to the user, if anything is added later, you can change it from DefaultOptions to something different.
> 
> Using `type RemoteOptions DefaultOptions`, it's like a symlink pretty much.
> 
> > Also, do you mean `type RemotesOptions DefaultOptions`?
> 
> yes, oops

Yeah, SGTM! Let's do this in another PR.

---

As for this PR, I think once we bring back the `RemoteSetURLAdd`, we're good to go!

jtagcat commented:
> Add does not take any regex.

See https://github.com/gogs/git-module/pull/67#discussion_r814613100

![image](https://user-images.githubusercontent.com/38327267/155693838-b3d57b7e-41ae-451b-8a3d-64c760113931.png)

Either adding the comment, or actually doing static checking would be nice.

jtagcat commented:
> As for this PR, I think once we bring back the RemoteSetURLAdd, we're good to go!

okie then

jtagcat commented:
LGTM

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,202 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+	Timeout time.Duration
+}
+
+// RemoteSetURLAdd appends an URL to the remote with given name of the repository in
+// given path. (RemoteSetURL overwrites the URL(s) instead)
```
```suggestion
// RemoteSetURLAdd appends an URL to the remote with given name of the repository in
// given path. Use RemoteSetURL to overwrite the URL(s) instead.
```

unknwon commented on repo_remote.go:
```diff
@@ -146,3 +146,202 @@ func RepoRemoveRemote(repoPath, name string, opts ...RemoveRemoteOptions) error
+}
+
+// RemoteSetURLAdd appends an URL to the remote with given name of the repository.
+// (RemoteSetURL overwrites the URL(s) instead)
```
```suggestion
// Use RemoteSetURL to overwrite the URL(s) instead.
```

unknwon commented:
Let's wait for the CI!

unknwon commented:
https://github.com/gogs/git-module/releases/tag/v1.2.0 has been tagged for this merge.
